tests/test_server.py

```diff
@@ -52,140 +71,152 @@ def test_add_device(client, name='steelseries-rival310'):
 
-    id = test_add_device(client, 'generic-hidpp20')
-    sleep(0.1) # Give time for the event nodes to be created
+    id = add_device(client, 'generic-hidpp20')
+    input_nodes = []
 
-    DeviceHandler.devices[id].verbose = False
+    start = time()
+    while len(input_nodes) == 0:
+        input_nodes = client.get('/devices/{}'.format(id)).json()['input_nodes']
+        sleep(0.1)
+        if time() > start + 3:
+            break
 
     # Claim the event nodes
     event_nodes = []
-    for node in DeviceHandler.devices[id].device_nodes:
+    for node in set(input_nodes):
         fd = open(node, 'rb')
         fcntl.fcntl(fd, fcntl.F_SETFL, os.O_NONBLOCK)
-        device = libevdev.Device(fd)
-        try:
-            device.grab()
-        except libevdev.device.DeviceGrabError:
-            with pytest.warns(UserWarning):
-                warnings.warn("Couldn't grab one of the event nodes", UserWarning)
-        event_nodes.append(device)
+        event_nodes.append(libevdev.Device(fd))
+
+    # make this exit on a timeout
+    from threading import RLock
+
+    events = []
+    def collect_events(stop):
+        nonlocal events
+        while not stop.is_set():
+            for node in event_nodes:
+                events += list(node.events())
+
+    import threading
+    stop_event_thread = threading.Event()
+    event_thread = threading.Thread(target=collect_events, args=(stop_event_thread,))
+    event_thread.start()
 
     # Send event
-    data = {
-        'x': 127,
-        'y': 127
-    }
-    response = client.post('/devices/{}/raw_event'.format(id),
-                           data=json.dumps(data),
-                           content_type='application/json')
+    data = [
+        {
+            'start': 0,
+            'end': 500,
+            'action': {
+                'type': 'xy',
+                'x': 5,
+                'y': 5
+            }
+        }
+    ]
+    response = client.post('/devices/{}/event'.format(id),
+                           json=data)
     assert response.status_code == 200
 
-    # Check if we received the event
-    x = y = 0
+    sleep(1)
+    stop_event_thread.set()
+    event_thread.join()
+
     for node in event_nodes:
-        events = node.events()
-        for e in events:
-            if e.matches(libevdev.EV_REL.REL_X):
-                x += e.value
-            elif e.matches(libevdev.EV_REL.REL_Y):
-                y += e.value
         node.fd.close()
-    assert x == data['x']
-    assert y == data['y']
 
+    m = MouseData()
+    for e in events:
+        if e.matches(libevdev.EV_REL.REL_X):
+            m.x += e.value
+        elif e.matches(libevdev.EV_REL.REL_Y):
+            m.y += e.value
+
+    dpi = response = client.get('/devices/{}/phys_props/dpi/{}'.format(id, 'active')).json()
+    assert m.x == int(data[0]['action']['x'] * 0.0393700787 * dpi)
+    assert m.y == int(data[0]['action']['y'] * 0.0393700787 * dpi)
+
+    client.delete('/devices/{}'.format(id))
```

Comment:
good, so here you just tested the default dpi.
Now we need to test various other dpis, the buttons, the keys, etc... :)